src/services/safe_border.py

```diff
@@ -0,0 +1,164 @@
+from src.web3py.typings import Web3
+from src.typings import BlockStamp
+from src.web3py.extentions.lido_validators import LidoValidator
+from src.modules.submodules.consensus import ChainConfig
+
+DEFAULT_SHIFT = 8 # epochs ~50 min
+MAX_NEGATIVE_REBASE_SHIFT = 1536 # epochs ~6.8 days
+FAR_FUTURE_EPOCH = 2**64 - 1
+MIN_VALIDATOR_WITHDRAWABILITY_DELAY = 2**8
+EPOCHS_PER_SLASHINGS_VECTOR = 2**13
+
+class SafeBorder:
+    def __init__(self, w3: Web3) -> None:
+        self.w3 = w3
+        self.lido_contracts = w3.lido_contracts
+
+    def get_safe_border_epoch(self, is_bunker: bool, blockstamp: BlockStamp, chain_config: ChainConfig):
+        if not is_bunker:
+            return self._get_new_requests_border_epoch(blockstamp)
+
+        self.chain_config = chain_config
+        
+        negative_rebase_border_epoch = self._get_negative_rebase_border_epoch(blockstamp)
+        associated_slashings_border_epoch = self._get_associated_slashings_border_epoch(blockstamp)
+
+        return min(
+            negative_rebase_border_epoch,
+            associated_slashings_border_epoch
+        )
+
+    def _get_new_requests_border_epoch(self, blockstamp: BlockStamp):
+        return self.get_epoch_by_slot(blockstamp.ref_slot) - DEFAULT_SHIFT
+
+    def _get_negative_rebase_border_epoch(self, blockstamp: BlockStamp):  
+        bunker_start_timestamp = self._get_bunker_mode_start_timestamp(blockstamp)
+        bunker_start_slot = self.get_slot_by_timestamp(bunker_start_timestamp)
+
+        last_withdrawable_epoch = self.get_epoch_by_slot(bunker_start_slot)
+
+        if bunker_start_slot > blockstamp.ref_slot:
```

Comment:
How it could be True?

Reply:
https://github.com/lidofinance/lido-dao/blob/feature/shapella-upgrade/contracts/0.8.9/WithdrawalQueue.sol#L41